lib/daru/dataframe.rb

```diff
@@ -295,11 +295,11 @@ def initialize source, opts={}
           end
         when Hash
           create_vectors_index_with vectors, source
+          vectors_have_same_index = all_vectors_have_equal_indexes?(source)
```

Comment:
My second point is that the `all_vectors_have_equal_indexes?` method is expecting a Hash of key value pairs where the values are strictly `Daru::Vector` objects, since this is ensured by the fact that it is inside the `if` block.

The program still works inspite of `source` being a Hash of key value pairs with `Array` objects for values because of duck typing. The [line in question](https://github.com/inside-track/daru/blob/FIX-DupePerformance/lib/daru/dataframe.rb#L2415) passes without error because `Array` has a method called `#index` defined on it.

So you can move the `vectors_have_same_index = all_vectors_have_equal_indexes?(source)` line below the first `if` statement (i.e. `if all_daru_vectors_in_source? source`) so that it is ensured that `source` always has key-value pairs where the values are `Daru::Vector` objects.

Also, I just reinstalled daru 0.1.2 from rubygems and am getting the same result. But I see your point and can merge this with the above changes.